feat(rocfft): integrate RCCL for single-node multi-GPU global transposition.

af-ayala opened the pull request:
## Motivation

Single-node multi-GPU FFT operations in rocFFT currently rely on peer-to-peer (P2P) hipMemcpyPeerAsync calls for data transfers during global transpose operations. While P2P works well for 2-GPU configurations, performance degrades at higher GPU counts (4+) due to lack of topology awareness and suboptimal routing through the interconnect fabric.

RCCL provides highly optimized collective operations that leverage the underlying GPU interconnect topology. By integrating RCCL for multi-GPU communication, rocFFT can achieve substantially better scaling on systems with 4+ GPUs, with benchmarks showing considerable speedups.

Jira ID: AIFFT-20

## Technical Details

* Added `rccl_wrapper.h/.cpp` providing a singleton communicator that initializes on first multi-GPU plan creation via `ncclCommInitAll`. The wrapper supports both uniform `ncclAllToAll` for symmetric transfers and grouped `ncclSend`/`ncclRecv` for non-uniform patterns, with complex data types handled by treating them as 2× real elements.

* Modified `plan.cpp` to call `GlobalTransposeRCCL()` when RCCL is available, which batches all cross-device transfers into a single RCCL group operation. New `CommRCCLAllToAll` and `CommRCCLGrouped` execution nodes in `tree_node.h/.cpp` manage the async launch and synchronization. The build system conditionally links RCCL via `-DROCFFT_RCCL_ENABLE=ON`.

* Added `comm_selector.h` with heuristics that auto-select P2P for ≤3 GPUs (where it's faster) and RCCL for 4+ GPUs when data exceeds threshold sizes. Users can override via `ROCFFT_COMM_BACKEND=P2P|RCCL` or disable RCCL entirely with `ROCFFT_DISABLE_RCCL=1`. The implementation gracefully falls back to P2P if RCCL initialization fails.

## Test Plan

Run full default CI,
Enable RCCL flag, build rocfft-test and run CI.

## Test Result

CI tests have passed.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

af-ayala committed:
RCCL INTEGRATION

malcolmroberts commented:
This needs a changelog entry.

evetsso committed:
default ROCFFT_RCCL_ENABLE to ON

evetsso committed:
build fixes when rccl is disabled

evetsso committed:
ROCFFT_RCCL_ENABLED -> ROCFFT_RCCL_ENABLE for consistency everywhere

evetsso committed:
RCCLCommunicator has setup/cleanup lifetime

evetsso committed:
RCCL prefix is redundant inside rocfft_rccl namespace

evetsso committed:
remove unused methods

af-ayala committed:
Merge branch 'develop' into users/alaayala/rccl_integration

malcolmroberts commented:
From the discussion today, the singleton pattern is problematic for the case where the user has more than one rccl-based transform per library instance.  We would need some information on how costly rccl initialization is, because this needs to happen at the plan generation level.

evetsso committed:
move rccl communicator to be owned by the plan

- it's a std::optional, so we get an exception instead of a crash if we
  try to use one that's not there
- Communicator can now assume pimpl is always there
- pass rccl communicator to CommRCCL classes
- move ops into communicator members
- log trace instead of cerr

evetsso committed:
ROCFFT_RCCL_ENABLE=OFF by default

evetsso committed:
make library-level world communicator, split it for smaller plans

evetsso committed:
don't split

evetsso committed:
remove extra stuff now that rank == device

evetsso committed:
fix norccl build

af-ayala committed:
Add entry to CHANGELOG

af-ayala committed:
Add entry to CHANGELOG

evetsso committed:
Merge remote-tracking branch 'upstream/develop' into users/alaayala/rccl_integration

evetsso committed:
format

evetsso committed:
remove stray conflict marker

regan-amd commented on projects/rocfft/library/src/include/plan.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2016 - 2023 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (C) 2016 - 2025 Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (C) 2016 - 2026 Advanced Micro Devices, Inc. All rights reserved.
```

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2680,6 +2861,32 @@ bool rocfft_plan_t::BuildOptMultiDevicePlan()
+            rccl = rocfft_rccl::Communicator::create(device_set);
+        }
+    }
+#endif
```
Such initializations of structure member (`this->rccl` in this case) should not be made specific to a given code path: this is prone to introducing code duplication in the future (_e.g._, if/when another new code path is added) and/or create unexpected object states in some scopes compared to others.

I'm suggesting to make this a step of `rocfft_plan_description_t::finalize_for(...)` if/when #4072 is merged.

regan-amd replied:
I would appreciate it if we could
- make `rccl` a member of `rocfft_plan_description_t`;
- make this part of `rocfft_plan_description_t::finalize_and_validate_for` (so that rccl is initialized for any code path that may be chosen thereafter);
- make sure the current device ID is also added to the set of device IDs.

af-ayala replied:
#4072 was closed, but still these suggestions have been taken into consideration. We still have not moved rccl as a member of rocfft_plan_description_t; but it can be in another PR.

regan-amd replied:
#4072 was indeed closed, but its core principles were not _abandoned_, just divided in smaller pieces to ease review and integration. The comment above from 05/05 meant to clarify the original request in light of the state of develop which had been updated since the inception of this work.

Global transposes are also used in the generalized multi-device code path introduced in #6691 but, as it currently stands, any transform committing to that new path _may or may not_ use RCCL: It depends on whether this pre-existing path was abandoned before or after reaching this point...
Such an uncertainty about RCCL being used (and usable) or not for a given plan configuration is very undesirable and easily avoidable. Guaranteeing that `rccl` is set and configured as needed before code path selection would
- address that ambiguity;
- widen the scope of influence of these RCCL-enabling changes (thereby increase test coverage and add confidence in the suggested changes);
- bring desirable performance improvements to all transforms for which a multi-device computation path exists, including real transforms of interest that are now enabled outside this specific path.

The scope of `rccl` (and where we decide to initialize it) is a design choice on which we have full control _right now_ given that `rccl` needs to be introduced in this very effort. We should come to a logic-motivated agreement now and act accordingly, but this cannot stand in my opinion.
Making `rccl` a member of the plan description and initializing it in `rocfft_plan_description_t::finalize_and_validate_for()` based on the description's I/O fields once they've been validated seems to be the most natural choice to me. If there is a compelling argument against that, its creation should at the very least be enforced again in the generalized multi-device code path. While that alternative would alleviate the concerns mentioned above, I believe it'd be logically more sound to make `rccl` a member of the description as I'm anticipating that even the gather-scatter strategy could/would also gain from using RCCL integration eventually (with additional, future changes).

regan-amd replied:
This was the most serious concern in my latest review and I see it marked as resolved without any explanation, now... The reasons developed above are justified, legitimate (based on actual observations on my end), and thoroughly explained. AFAICT, the concerns still exist: please let me know if I'm missing something obvious or share your reasons for dismissing the comment.

af-ayala replied:
Addressed in https://github.com/ROCm/rocm-libraries/pull/3042/changes/7a8081e62a5670ce828c4462ac22d36b70f2d7c7

Thank you.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2680,6 +2861,32 @@ bool rocfft_plan_t::BuildOptMultiDevicePlan()
+#ifdef ROCFFT_RCCL_ENABLE
+    // initialize RCCL for single-process multi-GPU if not already initialized
+    if(rocfft_plan_description_t::multiple_devices_in_rank(desc.inFields.front())
+       || rocfft_plan_description_t::multiple_devices_in_rank(desc.outFields.front()))
```
I know this is not really of your doing but this condition is questionable: the branch would also be taken in case of multi-process configurations with multiple devices per processes. That's not considered in tests (and not clear if we want to actually support such usage in the end), but it would be safer to also check that fields have only one ranks and that the rank is the same in either field.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2067,6 +2083,171 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        }
+    }
+}
+#endif // ROCFFT_RCCL_ENABLE
```
Unless I'm mistaken, there is a fair bit of duplication with `GlobalTransposeP2P` with some mild differences, _e.g._, some same-device data movements avoid some packing/unpacking... We should
1. have identical logic steps between `GlobalTransposeP2P` and `GlobalTransposeRCCL` for _fair_ performance comparisons, _i.e._, if one avoids packing/unpacking when possible so should the other;
2. alleviate duplications for similar-purpose functions with mild differences.

Non-type templating, _e.g._,
```cpp
template<InProcessComm_t p2pComm>
rocfft_plan_t::InProcessGlobalTranspose(...)
```
where `InProcessComm_t` would be an enum class with 2 possible values (_e.g._, `rcclComm` or `hipComm`) would address these concerns, for instance.

regan-amd replied:
Also, once implementations are aligned and actually fair to be compared to one another, do we have any guarantee/knowledge that RCCL will always outperform the alternative? Should it really be selected no matter what?

af-ayala replied:
This will be part of another PR for performance tuning.

regan-amd commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -658,6 +658,188 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
 }
 
+#ifdef ROCFFT_RCCL_ENABLE
+// RCCL AllToAll implementation (currently unused - kept for potential future use)
```
If anything is unused, it should be commented out for now to
1. avoid misleading code coverage reports;
2. be reviewed and tested more carefully when actually used and uncommented.

af-ayala replied:
I have updated the file accordingly

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,363 @@
+// Copyright (C) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (C) 2026 Advanced Micro Devices, Inc. All rights reserved.
```

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,119 @@
+// Copyright (C) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (C) 2026 Advanced Micro Devices, Inc. All rights reserved.
```

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,119 @@
+#include <rccl/rccl.h>
+#endif
+
+namespace rocfft_rccl
```
What's the rationale for this namespace? If it's simply to avoid name collisions, I'd prefer to rename the two classes it contains (which introduce their own namespace) with some rccl-specific names.

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,119 @@
+#include <set>
+#include <vector>
+
+#ifdef ROCFFT_RCCL_ENABLE
```
How reasonable is it to include _any_ of this header if `ROCFFT_RCCL_ENABLE` is not defined by the including unit? I find it odd to have the below classes and its member functions declared (but not functional) if this file is included by a unit with `ROCFFT_RCCL_ENABLE` undefined. Shouldn't we guard the _whole_ file and simplify its content accordingly instead?

af-ayala replied:
Updated in latest version

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,119 @@
+
+#ifdef ROCFFT_RCCL_ENABLE
+        struct Impl;
+        std::unique_ptr<Impl> pimpl;
```
Wouldn't it make more sense to declare this one a `std::shared_ptr<Impl>` and make object creation (static) member functions return plain (default-copyable) `Communicator` objects instead of `std::shared_ptr<Communicator>` ones?

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2052,8 +2059,17 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        GlobalTransposeRCCL(
+            elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
+    }
+    else
```
```suggestion
            elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
            return;
    }
```

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2042,6 +2046,9 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
                                     std::vector<size_t>&       outputItems,
                                     size_t                     transposeNumber)
 {
+    bool use_p2p = rocfft_plan_description_t::multiple_devices_in_rank(inField)
```
```suggestion
    const bool use_p2p = rocfft_plan_description_t::multiple_devices_in_rank(inField)
```
or keep it in the only `if` statement where it is used like before (preferred, personally).

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,363 @@
+{
+    std::shared_ptr<Communicator> Communicator::comm_world;
+
+#ifdef ROCFFT_RCCL_ENABLE
```
cf. comment in corresponding header: how reasonable is it to have any of what this unit defines available if `ROCFFT_RCCL_ENABLE` is not defined? I feel like we should guard the whole file and simplify its content accordingly instead.

af-ayala replied:
Fixed

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,363 @@
+            return ncclFloat32;
+        case 8:
+            // complex<float> (8 bytes) - use float32, caller doubles count
+            return ncclFloat32;
```
So, if ever using on real field, in double precisions (`elem_size == 8`), this would return `ncclFloat32`. That might lead to nightmarish debugging sessions... If there is a better way to duct-tape this, I would appreciate it.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,363 @@
+
+        if(elem_size == 8)
+        {
+            // complex float
```
The function's signature should be redesigned to account for an array type (--> distinguishing between real/complex elements) and a precision (--> singlehandedly defining `dtype`). As of now, this is correct only for complex arrays; future development may require real arrays.

I don't see the point of `get_nccl_datatype` if `dtype` is overwritten thereafter anyways. I don't think there is any  usage with `elem_size < 8` or `elem_size > 16` at the moment.

af-ayala committed:
Merge branch 'develop' into users/alaayala/rccl_integration

af-ayala committed:
changing comm initialization

af-ayala committed:
replace ncclCommInitAll with per-rank ncclCommInitRank and fix ambiguous dtype mapping

af-ayala committed:
activate ncclAllToAll for uniform decompositions and add RCCL path debug logging

af-ayala committed:
Update geometry querying

af-ayala committed:
minor

af-ayala committed:
Implace alltoall comments and fixes

af-ayala committed:
Fix default nccl types

af-ayala committed:
scope devices in AllToAll::Wait()

af-ayala committed:
ensure thread safety on comm_world

af-ayala committed:
Build RCCL updates

af-ayala committed:
Minor fixes

af-ayala committed:
Adding comments

af-ayala committed:
Merge branch 'develop' into users/alaayala/rccl_integration

af-ayala committed:
Resolving conflicts

af-ayala committed:
Resolving conflicts

af-ayala committed:
fix ncclAllToAll data race - add RCCL backend switch

af-ayala committed:
apply reviewer feedback (file-level guard, rename, value pimpl, single-proc guard)

evetsso commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+        return false;
+    }
+
+    bool HasTransfers() const
```
Not like this is all that harmful to leave in, but nothing's currently using this method.

af-ayala replied:
Fixed, comment from Rapahel below requires this method now.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+            outputItems.push_back(unpackIdx);
+        }
+
+        if(!rcclGrouped->transfers.empty())
```
```suggestion
        if(rcclGrouped->HasTransfers())
```
\+ make `CommRCCLGrouped::transfers` private.

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+        size_t              count;
+        bool                is_send; // true=send, false=recv
+        hipStream_wrapper_t stream; // each transfer has its own stream
+    };
```
Could this structure be made private within `CommRCCLGrouped `?

af-ayala replied:
Yes, moved it as private.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+    // multiple threads does not race on comm_world.
+    std::lock_guard<std::mutex> lock(comm_world_mutex);
+
+    if(!comm_world)
```
Doesn't this implicitly assume that the (possibly many) threads will all re-use the same set of devices? If we consider two plans being created, and having a shared existence span, the first one using device IDs `{0,1,2}` and the second using device IDs `{3,4,5,6,7}`, wouldn't the second one be given a smaller and different `comm_world` than expected?
I suspect we can't store a single static instance of `rocfft_rccl_comm_t`, but we may need more than one, as several device sets may be used throughout an application's lifetime?

af-ayala replied:
That's a great point and indeed, we should address this case, in commit 2be9a020fabbe3ee5db0d95b160d1c54c6636872, I am proposing a map from device-set to communicator, so each distinct set of devices gets its own RCCL communicator.

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+#include <hip/hip_runtime.h>
+#include <memory>
+#include <set>
+#include <vector>
```
```suggestion
#include <vector>
#include <cstddef>
```
for `size_t`

af-ayala replied:
addressed.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+
+    // device_id -> NCCL rank mapping
+    std::map<int, int> device_to_rank;
+    int                nranks = 0;
```
I feel like `nranks` is unnecessary as it must be equal to `comms.size()` and `device_to_rank.size()` unless I'm misunderstanding something. Please clarify if I miss something; if not, I'd prefer not storing `nranks` separately as it adds risk of producing inconsistent object states.

af-ayala replied:
Agreed, updated.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+            {
+                if(hipSetDevice(dev) != hipSuccess)
+                    throw std::runtime_error("hipSetDevice failed for device "
+                                             + std::to_string(dev));
```
Any reason we're not using [`rocfft_scoped_device`](https://github.com/ROCm/rocm-libraries/blob/000162b80b9d7c522059605f0623e601473f702c/projects/hipfft/shared/rocfft_hip.h#L27) instead?

af-ayala replied:
Updated to use `rocfft_scoped_device`

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+        // use try/catch to guarantee ncclGroupEnd is called even if
+        // hipSetDevice throws between ncclGroupStart and ncclGroupEnd.
+        // ranks are assigned in sorted device-id order (std::set).
+        ncclGroupStart();
```
Why not use a `rocfft_rccl_group_t` instance instead?

af-ayala replied:
Updated to use rocfft_rccl_group_t

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+    ncclUniqueId uniqueId{};
+
+    // device_id -> NCCL rank mapping
+    std::map<int, int> device_to_rank;
```
Could we have a simple `std::map<int, ncclComm_t> device_to_comms;` and get rid of `std::vector<ncclComm_t> comms;` instead? If the rccl rank associated with a device id is needed/queried thereafter we can always get it via `ncclCommUserRank`.

af-ayala replied:
Thanks for the suggestion, it's implemented now.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+{
+    auto it = pimpl->device_to_rank.find(device_id);
+    if(it == pimpl->device_to_rank.end())
+        return -1;
```
Unless it is very expected to call this routine with device IDs that are not in the communicator, I'd prefer having an `std::invalid_argument` exception thrown here and clearly stating that the queried device ID does not belong to the communicator. (The comment also applies to `void* rocfft_rccl_comm_t::get_comm(int device_id) const` above)
If throwing is not an acceptable option, the hardcoded `-1` should at least be replaced by a compile-time constant declared in `rccl_wrapper.h`, and used accordingly in implementations as that's easier to maintain.

af-ayala replied:
updated, thanks!

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+        {
+            return {};
+        }
+        new_comm.pimpl->uniqueId = id;
```
```suggestion
        ncclResult_t result = ncclGetUniqueId(&new_comm.pimpl->uniqueId);
        if(result != ncclSuccess)
        {
            return {};
        }
```

af-ayala replied:
thank you!

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+                  int         device_id,
+                  hipStream_t stream,
+                  size_t      base_type_size,
+                  bool        is_complex);
```
Could we replace these two argument by `rocfft_array_type` and `rocfft_precision` arguments for true integration within the project? I feel like any usage would unavoidably require conversion from such values anyways; let's have the wrapper do the conversions so that doesn't get duplicated if used in more than one place at some point.
Same comment for `send` and `recv`.

af-ayala replied:
Fixed in commit: dd86c784b396208a1ae8b95304bd7d860376b66e

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+              int         device_id,
+              hipStream_t stream,
+              size_t      base_type_size,
+              bool        is_complex);
```
`alltoall`, `send`, and `recv` all return a `bool` as error code. In their current usage, whenever `false` is returned, an exception is constructed and thrown right away... Do we foresee any usage when returning `false` would result in something _else_ than throwing an exception?

If _not_, could we sign the member functions `void` and let them throw instead? A specific exception type, intrinsically carrying information in its type, could be used too to make it easier to handle this kind of error anywhere, _e.g._, 
```cpp
struct rocfft_rccl_exception_t : std::runtime_error
{
    using std::runtime_error::runtime_error;
}
```
defined within the public section of this class.

af-ayala replied:
thank you, done in 32297b5e894b8a8ac633e241bdf72168d9a32da1

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+        , recvBuffers(_recvBuffers)
+        , sendOffsets(_sendOffsets)
+        , recvOffsets(_recvOffsets)
+    {
```
Would it make sense to check that
- all vector sizes are equal, and equal to `rccl.size()`;
- _all_ elements in `_locations` have their `comm_rank` member equal to `local_comm_rank`;
- all elements in `_locations`  have device IDs that are actually present in the `rccl` communicator;

and throw if any of the above is not verified? We could simplify/get rid of all `if(loc.comm_rank == local_comm_rank)` internally, if this is guaranteed true for all `loc` at construction. That would simplify logic in several places. If there is a _need_ to accept `loc.comm_rank != local_comm_rank`, please clarify why.

af-ayala replied:
That makes sense. I have added suggested validations in this part.

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+    {
+    }
+
+    void AddTransfer(bool              is_send,
```
It looks like the `is_send` argument would always be known at compile time. Please correct me if that's not true; otherwise, hard-coded bool value arguments (_i.e._, explicit `true`/`false` in function calls) make code harder to read. Could we consider non-type templating instead? For instance, 
```cpp
enum class rccl_op
{
    send,
    recv
};

template <rccl_op transfer_kind>
void AddTransfer(int peer_rank, ....
```
(and replace `Tranfer::is_send` by a `rccl_op` variable too, then)

af-ayala replied:
This suggestion is very nice. Thanks!

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2883,6 +3255,27 @@ bool rocfft_plan_t::BuildOptMultiDevicePlan()
+        std::set<int> device_set;
+        for(const auto& brick : desc.inFields.front().bricks)
+        {
+            if(brick.location.comm_rank == local_comm_rank)
```
Can this condition ever be `false` given the above check on `desc.get_local_comm_size() == 1`?

af-ayala replied:
This has been addressed in ba5439226daabbecb8d0c390661bbdc960b96bd7

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2256,16 +2260,17 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+    {
+        GlobalTransposeRCCL(
+            elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
+        return;
```
Would it make sense to have this logic as
```cpp
if(desc.get_local_comm_size() == 1)
{
#ifdef ROCFFT_RCCL_ENABLE
    try
    {
            GlobalTransposeRCCL(
                elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
            return;
    }
    catch(...)
    {
            if(LOG_PLAN_ENABLED())
                *LogSingleton::GetInstance().GetPlanOS()
                    << "GlobalTransposeRCCL could not be used"<< std::endl;
    }
#endif
    GlobalTransposeP2P(
            elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
}
else
{
    GlobalTransposeA2A(
            elem_size, inField, outField, input, output, inputAntecedents, outputItems, itemGroup);
}
```
- to unify single-proc-multi-device selection logic with what is being introduced elsewhere;
- so that the regular P2P would still be attempted in case of failure in the attempt at RCCL, instead of failing to create a plan?

af-ayala replied:
Implemented in new commit.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+{
+    ncclDataType_t dtype;
+    size_t         count_multiplier; // 1 for real, 2 for complex
+};
```
The `count_multiplier` member is a very simple re-interpretation of a `bool` flag given to `get_nccl_type_info`: I feel like it's unjustified to introduce a structure to that end, since using directly `(is_complex ? 2 : 1) * ` wherever applicable would make the code much more readable imo.

af-ayala replied:
Addressed in 6fb650d49e76b607003489e9501869f3cd87ad07

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,314 @@
+// multiplier.  base_type_size is the size of one real component
+// (2 for half, 4 for float, 8 for double).  complex data is
+// transferred as 2x the real component type.
+static NcclTypeInfo get_nccl_type_info(size_t base_type_size, bool is_complex)
```
If other comments are accounted for, the `bool` argument could be taken out, and this could be re-defined as
```cpp
static ncclDataType_t get_real_nccl_type(rocfft_precision prec)
```
which would be self-explanatory.

af-ayala replied:
addressed following comment above.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+                                        const std::string&         itemGroup)
+{
+    const auto   local_comm_rank = desc.get_local_comm_rank();
+    const size_t ndevices        = inField.bricks.size();
```
```suggestion
    const size_t nbricks_in        = inField.bricks.size();
```
`ndevices` is misleading (on input, on output, or combined?) + it's possible to have several bricks on the same device.

af-ayala replied:
Updated accordingly in all necessary places.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        size_t        inBrickIdx;
+        size_t        outBrickIdx;
+        data_layout_t intersection;
+        size_t        count;
```
I suspect `count` is not needed: `intersection.logical_count()` can be used instead as it returns that value at practically no cost.

af-ayala replied:
Updated following this comment. Thanks

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        size_t        outBrickIdx;
+        data_layout_t intersection;
+        size_t        count;
+        bool          same_device;
```
Same for this one, I'd rather see `inField.bricks[inBrickIdx].location == outField.bricks[outBrickIdx].location` wherever relevant, as that's not supposed to be prohibitively expensive.

af-ayala replied:
Updated following this comment. Thanks

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+                continue;
+
+            bool same_dev = (inBrick.location.device == outBrick.location.device
+                             && inBrick.location.comm_rank == outBrick.location.comm_rank);
```
```suggestion
            bool same_dev = (inBrick.location == outBrick.location);
```

af-ayala replied:
Added. Thanks for the suggestion

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+    // ranks in the communicator to participate
+    if(cross_device_count != ndevices * (ndevices - 1))
+        alltoall_eligible = false;
+    if(alltoall_eligible && rccl && static_cast<size_t>(rccl.num_ranks()) != ndevices)
```
```suggestion
    if(alltoall_eligible && static_cast<size_t>(rccl.num_ranks()) != ndevices)
```
can we do anything at all without `rccl` in here? Isn't that a requirement to even enter this function's scope?

af-ayala replied:
addressed

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+    void* get_comm(int device_id) const;
+
+    // total number of ranks in this communicator
+    int num_ranks() const;
```
I believe we always expect strictly positive values, right? If yes, let's return `size_t` to avoid the need to introduce a bunch of casts.

af-ayala replied:
addressed in latest commits.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        Grouped,
+    };
+    static const RcclBackend backend_choice = []() {
+        const auto v = rocfft_getenv("ROCFFT_RCCL_BACKEND");
```
```suggestion
        const auto v = std::to_lower(rocfft_getenv("ROCFFT_RCCL_BACKEND"));
```
so that `ROCFFT_RCCL_BACKEND=GROUPED`, `ROCFFT_RCCL_BACKEND=Grouped`, and `ROCFFT_RCCL_BACKEND=grouped` all get the same treatment...

af-ayala replied:
Thank you, updated with `std::tolower`

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        alltoall_eligible = false;
+    if(alltoall_eligible && rccl && static_cast<size_t>(rccl.num_ranks()) != ndevices)
+        alltoall_eligible = false;
+
```
I doubt the current logic for setting `alltoall_eligible` is robust enough. For instance, given that `ndevices` is actually the number of (input) bricks in here, it might be worth verifying that all input bricks are indeed on different devices, that all output bricks are also indeed on different devices as well and that the set of devices used on input matches the set of devices used on output.

Unless I'm mistaken, if
- `inField` has 2 bricks on device IDs `0` and `1`, first brick has full data but second brick is empty,
- `outField` has 2 bricks both on device ID `1`, both with half the data,

then I believe `alltoall_eligible` would be `true` and I believe that's incorrect.

af-ayala replied:
Thank you, I've improved the eligibility checks in bbe397bcb40e41cdc433fcbd1adc03d4431efb9f

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        {
+            const char* reason
+                = (backend_choice == RcclBackend::Grouped)
+                      ? (alltoall_eligible ? ", forced via ROCFFT_RCCL_BACKEND=grouped" : "")
```
```suggestion
                      ? (alltoall_eligible ? ", forced via ROCFFT_RCCL_BACKEND" : "")
```

af-ayala replied:
Suggestion applied, thank you,

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+                {inputAntecedents[info.inBrickIdx]});
+            multiPlan[packIdx]->group = itemGroup;
+            packItems.push_back(packIdx);
+        }
```
I feel like the above would gain (a lot) in clarity if `a2aSendBufs` and `a2aRecvBufs` were defined slightly differently, such that `a2aSendBufs[r]` and `a2aRecvBufs[r]` correspond to rccl rank `r` for all `0 <= r < rccl->num_rank()`. There would be no need for the `device_to_brick` map which maps device IDs to _input_ brick index yet is also being used with output bricks. Instead, for every `brick` of interest, `rccl->get_rank(brick.location.device)` could then be used and that would make things significantly easier to follow imo...

af-ayala replied:
I have made a re-structure based on this comment.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+            recvBuffers.push_back(BufferPtr::temp(a2aRecvBufs[d].data()));
+            sendOffsets.push_back(0);
+            recvOffsets.push_back(0);
+        }
```
Couldn't we have worked directly with `sendBuffers` and `recvBuffers` instead, and not use the intermediary `a2aSendBufs` and `a2aRecvBufs` at all?

af-ayala replied:
This was addressed following an update from another comment above.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        for(auto& buf : a2aSendBufs)
+            packBufs.push_back(std::move(buf));
+        for(auto& buf : a2aRecvBufs)
+            packBufs.push_back(std::move(buf));
```
What are the above needed for? `packBufs` goes out of scope immediately after these instructions: what is actually "kept alive" and how?

af-ayala replied:
Addressed in 5b84bd3d38fd05ddd3bdd8098b98960617a8d7dc

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+        rcclGrouped->description = "RCCL grouped send/recv";
+
+        std::vector<size_t> packItems;
+        packBufs.reserve(packBufs.size() + 2 * cross_device_count);
```
I believe `packBufs` can be declared local to this scope...

af-ayala replied:
resolved by other recent commits

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+            packItems.push_back(packIdx);
+
+            const int dst_rank = outBrick.location.device;
+            if(dst_rank >= 0)
```
```suggestion
            const int dst_rank = rccl->get_rank(outBrick.location.device);
            if(dst_rank >= 0)
```
or am I missing something? (same below for `src_rank`)

Also, would `src_rank` or `dst_rank` ever be (expected to be) negative? Shouldn't that be considered an unrecoverable logic error instead and trigger an exception to be thrown?

af-ayala replied:
Fixed already

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+                if(multiPlanAntecedents[unpackIdx].empty())
+                {
+                    AddAntecedent(unpackIdx, rcclIdx);
+                }
```
```suggestion
                AddAntecedent(unpackIdx, rcclIdx);
```
The antecedent should be added regardless of whether there already exist some or not: unpacking items **must** wait for completion of the grouped sends & receives. Am I missing something?

af-ayala replied:
You're right. I've updated accordingly.

regan-amd commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -555,7 +555,9 @@ void CommPointToPoint::ExecuteAsync(const rocfft_plan                     plan,
-        log_plan("CommPointToPoint\n");
+        log_plan("CommPointToPoint: " + std::to_string(numElems) + " elems, src device "
+                 + std::to_string(srcLocation.device) + " -> dst device "
+                 + std::to_string(destLocation.device) + "\n");
```
```suggestion
        log_plan("CommPointToPoint: " + std::to_string(numElems) + " elems, src device "
                 + srcLocation.str() + " -> dst device "
                 + destLocation.str() + "\n");
```

af-ayala replied:
Added. Thank you.

regan-amd commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -658,6 +660,231 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
+        const auto& loc = locations[i];
+        if(loc.comm_rank == local_comm_rank)
+        {
+            rocfft_scoped_device dev(loc.device);
```
```suggestion
```
is it actually needed here?

af-ayala replied:
Already removed by a fix from another comment

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+    bool send(const void* sendbuf,
+              size_t      count,
+              int         peer_rank,
+              int         device_id,
```
Unless there is a strong constraint justifying this, I would prefer using the same kind of logical arguments for _**both**_ of these arguments above: either ranks or device IDs for _both_, not rank for one and device ID for the other.

Whichever is chosen, I also suggest renaming `peer_rank` and `device_id` into `receiver_{rank,device_id}` and `sender_{rank,device_id}` (and `stream` into `sender_stream` in this case) to make the declaration more self-explanatory. (Similar comment for the `recv` member function).

af-ayala replied:
fixed in latest commits.

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+    const std::vector<size_t>            recvOffsets;
+
+    // streams for async execution
+    std::vector<hipStream_wrapper_t> streams;
```
If several vectors are indeed needed and if we can/should enforce them of being the same size, I would suggest using a vector of a structures, _e.g._,
```cpp
struct rccl_a2a_agent_t
{
    rocfft_location_t location;
    BufferPtr sendBuffer, recvBuffer;
    size_t sendOffset, recvOffsets;
    hipStream_wrapper_t stream;
}
std::vector<rccl_a2a_agent_t> agents; // agents[r] associated with rank r within rccl communicator
```
to guarantee consistency by design.

af-ayala replied:
Updated. Thank you, this comes more naturally, even more after removing `_locations`

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+              hipStream_t stream,
+              size_t      base_type_size,
+              bool        is_complex);
+
```
could `rocfft_rccl_comm_t::alltoall`, `rocfft_rccl_comm_t::send` and `rocfft_rccl_comm_t::recv` be labeled `const`? I don't think they're expected/supposed to modify the calling object, are they?

af-ayala replied:
Updated in latest commits

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+    }
+
+private:
+    rocfft_rccl_comm_t& rccl;
```
```suggestion
    const rocfft_rccl_comm_t& rccl;
```
Is this structure expected/supposed to ever modify that? If yes, please clarify how and why.
Same comment for `CommRCCLGrouped`.

af-ayala replied:
Updated to be const.

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+    static rocfft_rccl_comm_t comm_world;
+
+    // get the RCCL communicator for a specific device
+    void* get_comm(int device_id) const;
```
Why do we return a `void*` here? Could a `const ncclComm_t&` be returned instead? (throwing if that cannot be done for any reason, _e.g._, invalid argument).

af-ayala replied:
Fixed in d96bbdc4ed

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+                     const std::vector<BufferPtr>&         _sendBuffers,
+                     const std::vector<BufferPtr>&         _recvBuffers,
+                     const std::vector<size_t>&            _sendOffsets,
+                     const std::vector<size_t>&            _recvOffsets)
```
It looks like these offsets are always uniformly `0` in the current usage. Unless we foresee a clear need for non-zero offsets in the near future, I would remove them from the current structure entirely and simplify it (at least until we can actually cover that somehow in testing).

af-ayala replied:
Removed.

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+                     size_t                                _count_per_rank,
+                     const std::vector<rocfft_location_t>& _locations,
+                     const std::vector<BufferPtr>&         _sendBuffers,
+                     const std::vector<BufferPtr>&         _recvBuffers,
```
Do we need `_locations`? Isn't that somehow duplicating the info already available within `_rccl`? Couldn't we work with only `_sendBuffers` and `_recvBuffers` (if ordering their content by corresponding rccl rank)?

af-ayala replied:
Great catch, thanks!, I updated here: 1c0e8abbf7c93d211b71e89a53dccc41eedd8f8b

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1214,6 +1218,183 @@ struct CommPointToPoint : public MultiPlanItem
+    struct Transfer
+    {
+        int                 peer_rank;
+        rocfft_location_t   local_location;
```
Same regarding these two members: let's make them both of the same logical kind, _i.e._, both ranks or both locations/device id unless there is a strong constraint forcing it this way.

af-ayala replied:
Done.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2274,13 +2279,380 @@ void rocfft_plan_t::GlobalTranspose(size_t                     elem_size,
+            outputItems.push_back(unpackIdx);
+        }
+
+        if(!rcclGrouped->transfers.empty())
```
```suggestion
        if(rcclGrouped->HasTransfers())
```
would allow for `CommRCCLGrouped::transfers` to be declared private.

af-ayala replied:
Already addressed.

regan-amd commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -658,6 +660,231 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
+            rocfft_scoped_device dev(loc.device);
+            if(streams[stream_idx] && hipStreamSynchronize(streams[stream_idx]) != hipSuccess)
+                throw std::runtime_error("hipStreamSynchronize failed for RCCL AllToAll on device "
+                                         + std::to_string(loc.device));
```
Could we use events instead to align with other `MultiPlanItem`s? Any reason stream synchronization was chosen instead?

af-ayala replied:
Addressed in 31c8498b60ef1a4c4924c5d7b073a6d53a72cf62.

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,130 @@
+                  int         device_id,
+                  hipStream_t stream,
+                  size_t      base_type_size,
+                  bool        is_complex);
```
Do we want to make it a possibility for this function to be called for some (_but not all_) device ids present in the communicator? I feel like it would always be called in a loop parsing all device IDs present in the communicator, given that this is is an all-to-all operation? If so, wouldn't it be simpler (and safer as that would prevent incorrect usage) to make this function take `std::vector<...>` arguments of size `get_rank()` for `senbuf`, `recvbuf` and `stream`.

af-ayala replied:
Yes, I'm enabling that in 18557815ef422bd93fa1430400ed4fee22c5c688

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+
+size_t rocfft_rccl_comm_t::num_ranks() const
+{
+    return pimpl->device_to_comm.size();
+}
+
+int rocfft_rccl_comm_t::get_rank(int device_id) const
+{
+    auto it = pimpl->device_to_comm.find(device_id);
+    if(it == pimpl->device_to_comm.end())
+        throw std::invalid_argument("rocfft_rccl_comm_t::get_rank: device_id "
+                                    + std::to_string(device_id)
+                                    + " is not part of this communicator");
+    int rank = -1;
+    if(ncclCommUserRank(it->second, &rank) != ncclSuccess)
+        throw std::runtime_error("rocfft_rccl_comm_t::get_rank: ncclCommUserRank failed");
+    return rank;
+}
+
+std::vector<int> rocfft_rccl_comm_t::get_devices() const
+{
+    // ranks are assigned in sorted device-id order in create(), so
+    // std::map's natural ordering already gives us devices in rank order.
+    std::vector<int> devices;
+    devices.reserve(pimpl->device_to_comm.size());
+    for(const auto& [dev, comm] : pimpl->device_to_comm)
+        devices.push_back(dev);
+    return devices;
+}
+
+// RAII group wrapper
+rocfft_rccl_group_t::rocfft_rccl_group_t()
+{
+    ncclGroupStart();
+}
+
+rocfft_rccl_group_t::~rocfft_rccl_group_t()
+{
+    ncclGroupEnd();
+}
```
I'm concerned about the `ncclResult_t` returned by `ncclGroupStart` and `ncclGroupEnd` being ignored altogether in the above. This would result in silent errors at execution possibly leading to incorrect results or hard-to-debug issues. I think that
- the constructor could and should throw if `ncclGroupStart` fails;
- a private internal flag (say `needs_ending`) should be set to `true` upon successful construction;
- we should introduce a `void rocfft_rccl_group_t::end()` member function that can throw, and we should use that one explicitly instead of destructors invoked upon exiting artificial scopes;
- `void rocfft_rccl_group_t::end()` should turn `needs_ending` to `false` upon its successful completion (and do nothing if `needs_ending` is false upon entry);
- the destructor should then look like
```cpp
rocfft_rccl_group_t::~rocfft_rccl_group_t() noexcept
{
    try
    {
        end();
    }
    catch(const rocfft_rccl_exception_t& e)
    {
        log_trace(__func__, "ncclGroupEnd failed in destructor", e.what());
    }
    catch(...)
    {
        log_trace(__func__, "ncclGroupEnd failed in destructor with unexpected exception");
    }
}
```
to alleviate duplications with `void rocfft_rccl_group_t::end()`.

af-ayala replied:
Addressed in new commit, thank you!

malcolmroberts commented on projects/rocfft/library/src/CMakeLists.txt:
```diff
@@ -422,6 +435,11 @@ if( ROCFFT_MPI_ENABLE )
 endif()
 
+# RCCL: activate RCCL code paths (no-op stubs when undefined).
+if( rccl_FOUND )
```
probably use ROCFFT_RCCL_ENABLE here instead.

af-ayala replied:
Updated, thank you.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2505,6 +2505,27 @@ std::vector<size_t> rocfft_plan_t::GlobalTranspose(const field_view_t&        in
+                *LogSingleton::GetInstance().GetPlanOS()
+                    << "GlobalTransposeRCCL could not be used, falling back to P2P/A2A: "
+                    << e.what() << std::endl;
+        }
```
```suggestion
        }
        catch(...)
        {
             if(LOG_PLAN_ENABLED())
                *LogSingleton::GetInstance().GetPlanOS()
                    << "Unexpected exception type thrown by GlobalTransposeRCCL, falling back to P2P/A2A"<< std::endl;
        }
```

af-ayala replied:
this was solved in another commit

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2528,6 +2549,380 @@ std::vector<size_t> rocfft_plan_t::GlobalTranspose(const field_view_t&        in
+        for(size_t r = 0; r < nranks; ++r)
+        {
+            const rocfft_location_t loc{local_comm_rank, rccl_devs_vec[r]};
+            a2aSendBufs.emplace_back(
+                tempBuffers, local_comm_rank, loc, nranks * uniform_count * elem_size);
+            a2aRecvBufs.emplace_back(
+                tempBuffers, local_comm_rank, loc, nranks * uniform_count * elem_size);
```
That may trigger some implementation headache but couldn't the size of these buffers be `(nranks - 1) * uniform_count * elem_size` (at least theoretically) given that local transposes are taken care of separately? This wouldn't necessarily need to be addressed now, it could be done in a follow up task if too cumbersome to implement right now.

af-ayala replied:
I'd rather defer to a follow-up PR since dropping the self-diagonal isn't feasible without abandoning ncclAllToAll (it needs a uniform nranks × count layout). It could be done later with an alltoallv-style approach.

regan-amd replied:
No issue there, I was merely verifying my own comprehension by running the idea by you.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2528,6 +2549,380 @@ std::vector<size_t> rocfft_plan_t::GlobalTranspose(const field_view_t&        in
+            rcclGrouped->AddTransfer<rccl_op::send>(outBrick.location,
+                                                    inBrick.location,
+                                                    BufferPtr::temp(pack.data()),
+                                                    0,
```
```suggestion
                                                    0 /* offset */,
```
 for readability (same below).

regan-amd commented on projects/rocfft/CHANGELOG.md:
```diff
@@ -16,6 +16,10 @@ Documentation for rocFFT is available at
 
 ## rocFFT 1.0.37 for ROCm 7.13
 
+### Added
+
+* Added optional RCCL (ROCm Collective Communications Library) backend for single-node multi-GPU communication, enabled via `-DROCFFT_RCCL_ENABLE=ON`.
+
```
This should be moved up.

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,158 @@
+// the macro undefined the file expands to nothing.
+#ifdef ROCFFT_RCCL_ENABLE
+
+#include <cstddef>
+#include <hip/hip_runtime.h>
+#include <memory>
+#include <set>
+#include <stdexcept>
+#include <vector>
+
+#include <rccl/rccl.h>
+#include <rocfft/rocfft.h>
+
+// thrown by rocfft_rccl_comm_t communication primitives when the
+// underlying RCCL call fails.  the distinct type lets callers
+// recognize and handle RCCL failures specifically while still
+// being catchable via std::runtime_error / std::exception.
+struct rocfft_rccl_exception_t : std::runtime_error
+{
+    using std::runtime_error::runtime_error;
+};
```
```cpp
struct rocfft_rccl_exception_t : std::runtime_error
{
    rocfft_rccl_exception_t(std::string message, ncclResult_t code)
        : std::runtime_error(message)
        , error(code)
    {
        what_message = std::move(message) + " (" + ncclGetErrorString(error) + ")";
    }

    const char* what() const noexcept override
    {
        return what_message.c_str();
    }

private:
    const ncclResult_t error;
    std::string  what_message;
};
```
Would simplify current (and future) usage and alleviate duplications in constructions.

af-ayala replied:
Fix applied, thank you.

regan-amd commented on projects/rocfft/library/src/include/tree_node.h:
```diff
@@ -1226,6 +1230,239 @@ struct CommPointToPoint : public MultiPlanItem
+                     rocfft_array_type         _arrayType,
+                     size_t                    _count_per_rank,
+                     std::vector<agent_t>      _agents)
+        : rccl(_rccl)
+        , precision(_precision)
+        , arrayType(_arrayType)
+        , count_per_rank(_count_per_rank)
+        , agents(std::move(_agents))
+    {
+        // validate caller-supplied agent count against the communicator
```
```suggestion
    {
        local_comm_rank = 0; // RCCL enabled for single-process only
        // validate caller-supplied agent count against the communicator
```
otherwise, I'm afraid `ExecutesOnRank` may misbehave.

af-ayala replied:
Added. Thank you!

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+        if(result != ncclSuccess)
+        {
+            return {};
+        }
```
Shouldn't we throw a `rocfft_rccl_exception_t` instead (and maybe catch it in the caller's scope if a backup path can/should be used then)? Alternatively, let's have something logged here (with the returned error code) so we can get some clue about what happened. Same comment applies below.

af-ayala replied:
Good catch! Fixed.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+        return it->second;
+    }
+
+    {
```
Why this scoping?

af-ayala replied:
fixed

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+                new_comm.pimpl->device_to_comm[dev] = comm;
+                ++rank;
+            }
+        }
```
I'd prefer using the suggested `rocfft_rccl_group_t::end()` in here (to avoid silent errors in critical creation steps), and remove this scoping.

af-ayala replied:
Fixed by addressing related comment above

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+                                    + " is not part of this communicator");
+    int rank = -1;
+    if(ncclCommUserRank(it->second, &rank) != ncclSuccess)
+        throw std::runtime_error("rocfft_rccl_comm_t::get_rank: ncclCommUserRank failed");
```
This could (and should) be an `rocfft_rccl_exception_t`

af-ayala replied:
I agree, fixed it.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+
+    // resolve precision/complex/device mapping once outside the loop
+    const auto devices    = get_devices();
+    const auto nccl_count = count * (array_type_is_complex(array_type) ? 2 : 1);
```
```suggestion
    const auto nccl_count = count * (array_type_is_interleaved(array_type) ? 2 : 1);
```

af-ayala replied:
Updated.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+                                              + ncclGetErrorString(result));
+            }
+        }
+    }
```
I'd prefer using the suggested rocfft_rccl_group_t::end() in here (to avoid silent errors in critical execution steps), and remove this scoping.

af-ayala replied:
Addressed

regan-amd replied:
I believe the artificial scoping is unnecessary, now. Let's remove it if you agree. (Same comment in `tree_node.cpp`)

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,306 @@
+
+            if(result != ncclSuccess)
+            {
+                log_trace(__func__, "ncclAllToAll failed", result);
```
I believe we can remove this now as that would be redundant with this recently-added general [logging](https://github.com/ROCm/rocm-libraries/blob/1d0c3f8b4a91e4997bd3850e3b3c4a92791ab9f2/projects/rocfft/library/src/include/rocfft_exception.h#L41) logic. Same comment applies to other instances below.

af-ayala replied:
Updated it as your suggestion, thank you!

regan-amd commented on projects/rocfft/library/src/include/rccl_wrapper.h:
```diff
@@ -0,0 +1,158 @@
+    void send(const void*       sendbuf,
+              size_t            count,
+              int               peer_rank,
+              int               device_id,
```
Is there too much difficulty in adapting logic for using device ID or ranks for _**both**_? If achievable, I believe this would make the interface a lot more self-consistent.

af-ayala replied:
Updated in commit: 3f96a9cb, now just keeping `peer_device_id`

regan-amd commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -660,6 +661,214 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
+
+void CommRCCLAllToAll::Print(rocfft_ostream& os, const int indent) const
+{
+    const std::string indentStr("    ", indent);
```
```suggestion
    const std::string indentStr(indent*4, ' ');
```
I believe values of `indent` larger than 4 would be UB in the original form, whereas the intent is to have `indent` times a tab-like indentation, right? Same comment below if confirmed.

af-ayala replied:
fixed, thank you.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,358 @@
+    if(result != ncclSuccess)
+    {
+        log_trace(__func__, "ncclGroupStart failed", result);
+        throw rocfft_rccl_exception_t("ncclGroupStart failed", result);
```
Note: tracing _here_ may result in (possibly misleading) duplications in tracing logs if the thrown exception lands in a catch block that reports it again, _e.g._, [here](https://github.com/ROCm/rocm-libraries/blob/79a4b6604e2a51937d6072b366fa89d6be9b4811/projects/rocfft/library/src/include/rocfft_exception.h#L41). My personal opinion is that the exception carries the necessary information, the logging should be done in catch blocks wherever it lands if/when considered relevant. I let you judge for yourself, depending on what behavior you actually want(ed) here.

malcolmroberts commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -660,6 +661,212 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
+    // device, so no rocfft_scoped_device is needed here (mirrors
+    // CommGather / CommScatter / CommPointToPoint).
+    for(auto& agent : agents)
+    {
```
This would be a great place to add ncclCommGetAsyncError to see if things have gone sideways.


https://rocm.docs.amd.com/projects/rccl/en/latest/api-reference/api-library.html#_CPPv421ncclCommGetAsyncError10ncclComm_tP12ncclResult_t

af-ayala replied:
added. thank you.

malcolmroberts commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,358 @@
+        return ncclFloat32;
+    case 8:
+        return ncclFloat64;
+    }
```
default case here would be wise.

af-ayala replied:
added, thank you.

af-ayala replied:
added, thank yoiu

malcolmroberts commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,358 @@
+{
+    // check if RCCL is disabled via environment variable
+    const char* disable_rccl = std::getenv("ROCFFT_DISABLE_RCCL");
+    if(disable_rccl && std::string(disable_rccl) == "1")
```
I think that we already have a feature flag here in that we can turn rccl support on and off via cmake.  I don't think that we should add an environment variable as well.

af-ayala replied:
thank you, updated.

malcolmroberts commented on projects/rocfft/library/src/tree_node.cpp:
```diff
@@ -660,6 +661,212 @@ void CommPointToPoint::Print(rocfft_ostream& os, const int indent) const
+        rocfft_scoped_device dev(devices[r]);
+        send_ptrs[r] = agents[r].sendBuffer.get(in_buffer, out_buffer, local_comm_rank, info);
+        recv_ptrs[r] = agents[r].recvBuffer.get(in_buffer, out_buffer, local_comm_rank, info);
+        streams[r]   = agents[r].stream;
```
If streams are provided in the execution info, shouldn't we use those streams for the alltoall communication?

regan-amd replied:
All of the pre-existing communication nodes have been using their own streams thus far, regardless of whether users gave an actual non-default stream or not, see
- [`CommPointToPoint::ExecuteAsync`](https://github.com/ROCm/rocm-libraries/blob/c5967e63981b555ff5efd4ab042ac1e5de7d847d/projects/rocfft/library/src/tree_node.cpp#L547);
- [`CommScatter::ExecuteAsync`](https://github.com/ROCm/rocm-libraries/blob/c5967e63981b555ff5efd4ab042ac1e5de7d847d/projects/rocfft/library/src/tree_node.cpp#L663);
- [`CommGather::ExecuteAsync`](https://github.com/ROCm/rocm-libraries/blob/c5967e63981b555ff5efd4ab042ac1e5de7d847d/projects/rocfft/library/src/tree_node.cpp#L795);
- [`CommAllToAll::ExecuteAsync`](https://github.com/ROCm/rocm-libraries/blob/c5967e63981b555ff5efd4ab042ac1e5de7d847d/projects/rocfft/library/src/tree_node.cpp#L930).

If you see that as an actual issue, it should be addressed consistently across all cases in my opinion, but breaking consistency may not be desirable at this point. My own understanding is that this was an _intentional_ implementation choice, motivated by the desire to enable possible concurrent executions (overlapping) of communication and independent computation steps, and/or to prevent serialization of independent tasks known to be "internal details" to the overall users' requests.

If indeed intentional, this may fail to honor implicit dependencies that would have been enqueued/expected in/from user-given stream(s) for the gather-scatter path though... I personally think that the honoring of user-given streams in multi-device code paths could/should take some revisions, but I do _not_ think that the best solution is to systematically use user-given stream(s), even for communications. In particular, these RCCL-enabling changes affect only use cases that are _not_ dispatched to the gather-scatter strategy, for now: for such use cases, I do think that such a change would be undesirable in fact, as that may unnecessarily hinder performance improvements.

eng-flavio-teixeira commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -2505,6 +2505,35 @@ std::vector<size_t> rocfft_plan_t::GlobalTranspose(const field_view_t&        in
+    {
+        try
+        {
+            return GlobalTransposeRCCL(input, output, antecedents);
```
If there is an exception here, it appears that GlobalTransposeRCCL can mutate multiPlan but I can't see any rollback mechanism when the RCCL global transpose fails. 

If that's the case, then the fallback P2P / A2A may have orphaned multiPlan items (from RCCL's transpose) executing with the fallback, possibly leading to incorrect results / doubling the amount of work to perform the global transpose.

af-ayala replied:
Thank you @eng-flavio-teixeira for this great catch, I've addressed it in 520e823ef0.

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -3764,6 +3764,36 @@ rocfft_plan_t::field_view_t rocfft_plan_t::field_view_t::get_embedding_view() co
+    {
+        if(brick.location.comm_rank == local_comm_rank)
+            device_set.insert(brick.location.device);
+    }
```
Please insert the current device to the set as well.

af-ayala replied:
Addressed, thank you!

regan-amd commented on projects/rocfft/library/src/plan.cpp:
```diff
@@ -4392,6 +4825,11 @@ static rocfft_status rocfft_plan_create_internal(rocfft_plan                   p
 
+#ifdef ROCFFT_RCCL_ENABLE
+        // init rccl before any plan-building path is chosen
+        plan->InitRCCLCommunicator();
```
If this throws an exception, we may want to log something but we could/should still be able to proceed without RCCL.

af-ayala replied:
Fixed. I wrapped the call in a try/catch that logs to the plan log and continues with an empty RCCL.

regan-amd commented on projects/rocfft/library/src/rccl_wrapper.cpp:
```diff
@@ -0,0 +1,366 @@
+            "ncclCommGetAsyncError failed on device " + std::to_string(device_id), result);
+    if(async_error != ncclSuccess)
+        throw rocfft_rccl_exception_t(
+            "RCCL asynchronous error on device " + std::to_string(device_id), async_error);
```
- I don't think `async_error == ncclInProgress` would justify throwing given the context in which this is being used, would it?
- Should `ncclCommAbort` be called before throwing in case of asynchronous error?

af-ayala replied:
Fixed in [0c0e2cb](https://github.com/ROCm/rocm-libraries/pull/3042/commits/0c0e2cbafce57994e3e6f6c8d0e2d187e6381bd3), thank you!
